### PR TITLE
VID: Reorganize and create new events

### PR DESCRIPTION
**Pull request by rob-maron**
Closes issue: #1817 

### This PR: 

- Creates a new event `TransactionsSequenced`
- Reorganizes events such that:
    - Transactions task - collects transactions and sequences them. Emits a TransactionsSequenced event that contains the payload and metadata
    - VID task - listens for TransactionsSequenced, calculates the encoded transactions, VID disperse data upon receiving that event. Emits BlockReady event, which contains the commitment and the disperse data.
    - VID task - listens for BlockReady, then sends out the VIDProposalSend event
    - DA task - listens for TransactionsSequenced, creates DA proposal upon receiving that event. Emits DAProposalSend
    
- Moves `SendPayloadCommitmentAndMetadata` to the VID task
- Change the tests to account for the new and moved events

Note that, in reference to the previous PR https://github.com/EspressoSystems/HotShot/pull/2077:
- I chose to pass around the payload and metadata as opposed to the encoded bytes, since the DA task needs access to the payload to create the DA proposal.

### This PR does not: 

- Touch the VID header or payload structure 
- Remove commitment computation (this is in #2102)

### Key places to review: 

The new logic as specified in `da.rs`, `vid.rs`, `transactions.rs`.

### CODE REVIEW

**Review by shenkeyao — CHANGES_REQUESTED** [review]
Some comments about the `TransactionsSequenced` event. Looks good overall!

File: crates/task-impls/src/da.rs [hunk 1]
```diff
@@ -371,11 +371,17 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
+            HotShotEvent::TransactionsSequenced(encoded_txns, metadata, view) => {
                 self.da_network
                     .inject_consensus_info(ConsensusIntentEvent::CancelPollForTransactions(*view))
                     .await;
 
+                // calculate payload from encoded transactions and metadata
+                let payload = <TYPES::BlockPayload as BlockPayload>::from_bytes(
+                    encoded_txns.into_iter(),
+                    metadata,
+                );
```
[the remark this answers is not in the document]
Comment (rob-maron) [reply crates/task-impls/src/da.rs:383]: That was the original plan, but per @shenkeyao 's suggestions I switched it back around. I should've updated the PR comment

File: crates/task-impls/src/da.rs [hunk 2]
```diff
@@ -158,6 +159,15 @@ async fn vote_handle<TYPES: NodeType, I: NodeImplementation<TYPES>>(
...
 }
 
+/// Helper function to calculate a sha256 hash
+/// Used primarily for signing and verifying the encoded transactions
+#[must_use]
+pub fn sha256_hash(data: &Vec<u8>) -> [u8; 32] {
+    let mut hasher = Sha256::new();
+    hasher.update(data);
+    hasher.finalize().into()
+}
```
Comment (ggutoski) [inline crates/task-impls/src/da.rs:169]: If you want you could also do
```rust
let hash = Sha256::digest(data);
```
See https://github.com/RustCrypto/hashes#usage

File: crates/task-impls/src/da.rs [hunk 3]
```diff
@@ -196,6 +206,7 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                 }
 
                 let payload_commitment = vid_commitment(proposal.data.encoded_transactions.clone());
```
Comment (ggutoski) [inline crates/task-impls/src/da.rs:208]: We don't want to clone the whole payload just to make a VID commitment. I'm surprised clippy doesn't complain.

File: crates/task-impls/src/transactions.rs [hunk 4]
```diff
@@ -220,53 +217,25 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
-                // TODO never clone a block
-                // https://github.com/EspressoSystems/HotShot/issues/1858
+
+                // send the sequenced transactions to VID and DA tasks
                 self.event_stream
-                    .publish(HotShotEvent::BlockReady(
+                    .publish(HotShotEvent::TransactionsSequenced(
                         encoded_transactions,
-                        metadata,
+                        metadata.clone(),
```
Comment (ggutoski) [inline crates/task-impls/src/transactions.rs:234]: I don't see why we need a clone here. Can't the `TransactionsSequenced` event just take ownership of `metadata`?

File: crates/task-impls/src/vid.rs [hunk 5]
```diff
@@ -336,6 +344,56 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
...
+                // send the block to the VID dispersal function
+                self.event_stream
+                    .publish(HotShotEvent::BlockReady(
+                        VidDisperse {
+                            view_number,
+                            payload_commitment: vid_disperse.commit,
+                            shares: vid_disperse.shares,
+                            common: vid_disperse.common,
+                        },
+                        vid_disperse.commit,
```
Comment (ggutoski) [inline crates/task-impls/src/vid.rs:374]: lol we have 2 copies of the VID commitment in this event 🙄

**Review by ggutoski — COMMENTED** [review]
Just a few minor comments.

**Review by ggutoski — APPROVED** [review]

**Review by shenkeyao — APPROVED** [review]
LGTM!

**Pull request merged**